### ISSUE TITLE
Sync Issue between Guide video and User Video Replay

### ISSUE THREAD

**Issue by datakalp** [datakalp/Vajra-Hands-App#18]
Recording made just now (5:14 pm, kalpit@datakalp.com)

I went to an area with weak wi-fi. Hence the guide video kept getting stuck while recording ...sometimes for several seconds. However the recording was always on even when the guide video was stuck. 
This caused two issues:
1) On the results page, it says the duration is 36 seconds, although the recording was for a much longer duration. 
2) The user video on the left became out-of-sync compared to the guide video shown on the right. This is bad user experience, because the user had only done motions in sync with the guide video. Also, this out of sync can result in low score even when user performed fully correct hand wash.

**Comment by datakalp** [comment]
Possible solutions:
1) Start loading the guide video as soon as user signs in, even before she clicks the record button. Keep that video in browser cache.
2) In case the guide video gets stuck, store the start and end timestamp of the guide video for each step. Then use these timestamps to align the guide video with the user video in the review screen as well as results screen.

**Comment by datakalp** [comment]
Looking at this, storing the guide video (a few MBs?) on the device shouldn't be a problem for PWA at all:
https://stackoverflow.com/questions/58591353/how-much-can-a-pwa-store

**Comment by anubhutiGupta** [comment]
Seems like single source playback (downloading video in one go before playing) is not possible; have tried compressing the guide video for smoother playback, let's see if we face this issue again

**Comment by datakalp** [comment]
Did we figure out why the App was still trying to fetch the video when it existed in the browser cache?

**Comment by anubhutiGupta** [comment]
From what we have learned. Unless our video file is very small, it’s unlikely that any video player will download the whole file in one go before playback.

**Comment by datakalp** [comment]
Ok...so the first time the user makes a recording, the entire guide video would be downloaded locally, right? Then 2nd time onwards can it be played from the local cache?

**Comment by anubhutiGupta** [comment]
Yes! it remains downloaded 100% if user attempts another recording that is ; the guide animation on the record screen has played once

**Comment by datakalp** [comment]
This issue will get closed once we are able to send accurate start | end times for the guide video steps from front end to backend.

**Comment by anubhutiGupta** [comment]
Taking this up now

**Comment by anubhutiGupta** [comment]
A few updates and observations on this issue.

1. This has not been reported any time soon because loading time of the animated guide video has been reduced significantly. We have achieved this by doing the following :- 
          - Compressing the animated guide video file
          - Streaming it from a public URL (Azure blob) ; which is considered as one of the good practices for increasing playback experience
          - Tested this on a slow 3G network ; and still video played fine without any delay          
2. The implementation of the solution ; that is mapping the video with the `gif_time` is achievable but critical; since animated guide will need to be paused and played again and again as per user's step 
3. So, if we do not face this issue any more we can consider implementing the solution as a low priority task, as of now

**Comment by datakalp** [comment]
Bumping up the priority as many issues reported in today's testing were related to the sync
